onnx/onnx_backend.h

```diff
@@ -0,0 +1,914 @@
+ *                                function completes with either
+ *                                ONNX_STATUS_SUCCESS or ONNX_STATUS_FALLBACK
+ *                                status codes, the actual size of the value
+ *                                queried in the call is stored in the variable
+ *                                specified by this pointer.
+ *
+ * @retval ONNX_STATUS_SUCCESS The function call succeeded, and requested value
+ *                             is stored in the location specified by infoValue,
+ *                             and the actual size of the requested value is
+ *                             stored in the location specified by
+ *                             infoValueSize.
+ * @retval ONNX_STATUS_FALLBACK The function call completed, but the requested
+ *                              value was not stored in the location specified
+ *                              by infoValue, either because it is NULL, or
+ *                              because the size of the memory buffer is
+ *                              insufficient for the value. The actual size of
+ *                              the requested value is stored in the location
+ *                              specified by infoValueSize.
+ * @retval ONNX_STATUS_INVALID_INDEX The function call failed because backend
+ *                                   index is out of bounds (is greater or equal
+ *                                   to the value returned by
+ *                                   onnxGetNumBackends)
+ * @retval ONNX_STATUS_INVALID_POINTER The function call failed because
+ *                                     infoValueSize is NULL.
+ * @retval ONNX_STATUS_UNSUPPORTED_PARAMETER The function call failed because
+ *                                           the value of infoType is
+ *                                           not supported by the backend.
+ */
+ONNX_PUBLIC onnxStatus ONNX_ABI ONNX_SYMBOL_NAME(onnxGetBackendInfo)(
+    uint32_t index,
+    onnxBackendInfo infoType,
+    void* infoValue,
+    size_t* infoValueSize);
+
+/**
+ * Query if an ONNX model graph is compatible with the backend.
+ *
+ * Model graph is passed as a serialized ModelProto message, where types and
+ * dimensions of all inputs (including static weights) and outputs are specified
+ * through ModelProto.graph.input and ModelProto.graph.output messages. If the
+ * backend supports ONNX_CAPABILITY_SYMBOLIC_SIZE_TENSORS, some of the shape
+ * dimensions can be symbolic. In this case, their validation should be deferred
+ * until graph inputs and outputs are specified in onnxSetGraphIO.
+ *
+ * Commonly, the serialized ModelProto message passed to this function would
+ * not include the static weights (ModelProto.graph.initializer is empty), and
+ * the backend implementation MUST NOT rely on the weights to determine if the
+ * graph is supported.
+ *
+ * An important use-case is a ModelProto containing only a single NodeProto in
```

Comment:
Why not have a separate function to check whether a single node in a given graph is compatible? Constructing a ModelProto for this case (which is what you'd want to do to check for partial compatibility) seems unnecessary.

Reply:
I think it's just from the consideration to make # of API functions smaller.

We could add the function for a single node - but then it needs to have access also to input/output shapes (as some backends might support only some sizes). GraphProto already conveniently has a place to store shape information

Also, at least from the use cases we're aware of, if the backend can support a graph then it can support any sub-part of the graph too (not sure whether it's true in the more general case).

Reply:
We considered three options: `NodeProto`, `GraphProto`, and `ModelProto`. However, when we tried to map existing vendor libraries to the backend API with the first two options, we realized that they lack important context details:

- `NodeProto` doesn't internally contain information about input and output shapes, and this information is important to decide is an operator is supported. Many of today's vendor libraries have a limit on maximum shapes that can handle, and we know one vendor library that imposes divisibility restrictions on number of input and output channels (i.e. number of channels must be a multiple of 8, otherwise operator is unsupported), thus input/output shapes are crucial for `onnxGetBackendCompatibility` operation.

- `GraphProto` is a better candidate as it provides input/output shapes and data types, but `GraphProto` still lacks crucial information to validate the graph: `opset_import` and `ir_version`.

Thus, we settled on `ModelProto` for the proposal. It has the additional advantage that much code implementing and using `onnxInitGraph` and `onnxGetBackendCompatibility` can be shared.

Reply:
May be we should have a single call where we can pass in a GraphProto and get back a list of all the nodes in the Graph that can be supported. This is useful if we want the partition the graph into subgraphs that can be run on the backend.

Reply:
We certainly should consider this option. It would reduce the time to discover the set of supported operations, but I have two concerns:

- The current `onnxGetBackendCompatibility` API can be implemented on top of all existing vendor libraries, e.g. by feeding the `ModelProto` to the library and checking if it fails. The suggested change would require new functionality in vendor libraries, and we need to check if vendors would be willing to implement it.
- With the current API, deep learning frameworks don't need to pay much attention to serialization of a model graph into ONNX format, and it is typically hidden behind a high-level function call. The proposed change would require deep learning frameworks to maintain mapping between the order of operators in serialized ONNX message and the same operators in their internal IR.

I think we'd need to do some experiments, and in particular quantify the win from this change before deciding on it.